datalad/support/dsconfig.py

```diff
@@ -0,0 +1,319 @@
+    ``.datalad/config`` inside a dataset too. This file is commit to Git and
+    hence useful to ship certain configuration items with a dataset.
+
+    The API aims to provide the most significant read-access bits of a
+    dictionary, the Python ConfigParser, and GitPython's config parser
+    implementation.
+
+    This class is presently not capable of effienciently writing multiple
+    configurations items at once. Instead, each modification results in a
+    dedicated call to `git config`. This authors thinks this is OK, as he
+    cannot think of a situation where a large number of items need to be
+    written during normal operation. If such need arises, various solutions are
+    possible (via GitPython, or an independent writer).
+
+    Parameters
+    ----------
+    dataset : Dataset, optional
+      If provided, all `git config` calls are executed in this dataset's
+      directory. Moreover, any modifications are, by default, directed to
+      this dataset's configuration file (which will be created on demand)
+    dataset_only : bool
+      If True, configuration items are only read from a datasets persistent
+      configuration file, if any (the one in .datalad/config, not .git/config).
+    """
+    def __init__(self, dataset=None, dataset_only=False):
+        # store in a simple dict
+        # no subclassing, because we want to be largely read-only, and implement
+        # config writing separately
+        self._store = {}
+        self._dataset = dataset
+        self._dataset_only = dataset_only
+        if dataset is not None:
+            # make sure we run the git config calls in the dataset
+            # to pick up the right config files
+            self._runner = Runner(cwd=dataset.path)
+        else:
+            self._runner = Runner()
+        self.reload()
+
+    def reload(self):
+        """Reload all configuration items from the configured sources"""
+        self._store = {}
+        # 2-step strategy: load git config from all supported sources
+        # then load datalad dataset config from dataset
+        # in doing so we always stay compatible with where Git gets its
+        # config from
+        if not self._dataset_only:
+            stdout, stderr = self._runner.run(
+                ['git', 'config', '-l'], log_stderr=False)
+            self._store = _parse_gitconfig_dump(
+                stdout, self._store, replace=False)
+        if self._dataset:
+            # now any dataset config
+            dscfg_fname = opj(self._dataset.path, '.datalad', 'config')
+            if exists(dscfg_fname):
+                stdout, stderr = self._runner.run(
+                    ['git', 'config', '-l', '--file', dscfg_fname],
+                    log_stderr=False)
+                # overwrite existing value, do not amend to get multi-line
+                # values
+                self._store = _parse_gitconfig_dump(
+                    stdout, self._store, replace=True)
```

Comment:
so `.datalad/config` would be overriding any setting -- global or local... kinda suboptimal... I might even say don't we want dataset config to be the first one considered and then possibly augmented by the non-persistent global and/or local?